PhysicsTools/NanoAOD/plugins/BToKsteeProducer.cc

```diff
@@ -0,0 +1,790 @@
+
+                    pair<double,double> DCA = computeDCA(kaon,
+							 bFieldHandle,
+							 beamSpot);
+                    double DCABS = DCA.first;
+                    double DCABSErr = DCA.second;
+
+                    if(fabs(DCABS/DCABSErr)<DCASigMinKaon_) continue;
+
+		    for (unsigned int l = 0; l < pfCandNumber; ++l) {
+
+		      if(k==l) continue;
+
+		      const pat::PackedCandidate & pion = (*pfCandHandle)[l];
+		      if(abs(pion.pdgId())!=211) continue; //Charged hadrons
+		      if(!pion.hasTrackDetails()) continue;
+		      if(pion.pt()<ptMinPion_ || abs(pion.eta())>etaMaxPion_) continue;
+		      if(KstCharge_ && kaon.charge()*pion.charge()>0) continue;
+
+		      pair<double,double> DCA = computeDCA(pion,
+							   bFieldHandle,
+							   beamSpot);
+		      double DCABS = DCA.first;
+		      double DCABSErr = DCA.second;
+		      
+		      if(fabs(DCABS/DCABSErr)<DCASigMinPion_) continue;
+
+		      bool passed_KstRefit = false;
+
+		      double KstLSBS = -1.;
+		      double KstLSBSErr = -1.;
+		      double KstVtx_CL = -1.;
+		      double Kst_mass_err = -1.;
+
+		      RefCountedKinematicParticle refitKst;              
+		      math::XYZVector refitKstV3D;
+
+		      if(save2TrkRefit_){
+
+			RefCountedKinematicVertex refitVertexKst;
+			RefCountedKinematicParticle refitKaon_Kst;
+			RefCountedKinematicParticle refitPion_Kst;
+
+			passed_KstRefit = KstVertexRefitting(kaon, pion,
+							     theTTBuilder,
+							     refitVertexKst,
+							     refitKst,
+							     refitKaon_Kst,
+							     refitPion_Kst);
+
+			if(passed_KstRefit){			  
+
+			  pair<double,double> KstLS = computeLS(refitVertexKst,beamSpot);
+			  KstLSBS = KstLS.first;
+			  KstLSBSErr = KstLS.second;
+
+			  KstVtx_CL = TMath::Prob((double)refitVertexKst->chiSquared(),
+						  int(rint(refitVertexKst->degreesOfFreedom())));
+
+			  Kst_mass_err = sqrt(refitKst->currentState().kinematicParametersError().matrix()(6,6));
+
+			  math::XYZVector refitKaonV3D_Kst = refitKaon_Kst->refittedTransientTrack().track().momentum();
+			  math::XYZVector refitPionV3D_Kst = refitPion_Kst->refittedTransientTrack().track().momentum();
+			  refitKstV3D = refitKaonV3D_Kst + refitPionV3D_Kst;
+
+			}
+
+		      }                    
+		      
+		      RefCountedKinematicVertex refitVertexBToKstEE;
+		      RefCountedKinematicParticle refitBToKstEE;
+		      RefCountedKinematicParticle refitEle1;
+		      RefCountedKinematicParticle refitEle2;
+		      RefCountedKinematicParticle refitKaon;
+		      RefCountedKinematicParticle refitPion;                    
+                    
+		      bool passed = BToKstEEVertexRefitting(ele1, ele2, kaon, pion,
+							    theTTBuilder,
+							    refitVertexBToKstEE,
+							    refitBToKstEE,
+							    refitEle1,
+							    refitEle2,
+							    refitKaon,
+							    refitPion);
+                    
+		      if (!passed) continue;
```

Comment:
for the Kstar channel, I would change the default logic
- for electrons it is ok as it is (it reflects what's done for the K case), without the constraint on the ee vtx in the final fit, because the final analysis selects JPsi wrt non-resonant based on the ee mass and we don't want bias
- for the K*, I think that the final vtx refit should be done instead with "refitKst" and the ele1 ele2.
Because looking for a K* state does not bias the selection of the resonant wrt non-resonant. (as done for the D* (Kpi) pi

So:
1) I would add a "bool passed = BToKstEEVertexRefitting(ele1, ele2, refitKst,..." that is anyway missing

2) For consistency with the ee part, I wouldn't skip the cases where the 4 (K pi e e)independent vtxFit is not ok. Worth to save the info, but I would remove the "continue" and change accordingly  the logic